nexus/db-queries/src/db/datastore/sled.rs

```diff
@@ -189,13 +216,43 @@ impl DataStore {
         resources: db::model::Resources,
         constraints: db::model::SledReservationConstraints,
     ) -> CreateResult<db::model::SledResource> {
-        #[derive(Debug)]
-        enum SledReservationError {
-            NotFound,
-        }
+        self.sled_reservation_create_inner(
+            opctx,
+            resource_id,
+            resource_kind,
+            resources,
+            constraints,
+        )
+        .await
+        .map_err(|e| match e {
+            SledReservationTransactionError::Connection(e) => e,
+            SledReservationTransactionError::Diesel(e) => {
+                public_error_from_diesel(e, ErrorHandler::Server)
+            }
+            SledReservationTransactionError::Reservation(e) => match e {
+                SledReservationError::NotFound
+                | SledReservationError::TooManyAffinityConstraints
+                | SledReservationError::ConflictingAntiAndAffinityConstraints
+                | SledReservationError::RequiredAffinitySledNotValid => {
+                    return external::Error::insufficient_capacity(
+                        "No sleds can fit the requested instance",
+                        "No sled targets found that had enough \
+                                 capacity to fit the requested instance.",
+                    );
+                }
+            },
+        })
+    }
 
+    async fn sled_reservation_create_inner(
```

Comment:
FYI @hawkw , you asked about this in the Hypervisor Sync.

The original API of the function `pub async fn sled_reservation_create` still exists above, but I made this "inner" function to get better testing of errors.

But we could also report these errors through e.g. an alerting system or something.